File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.py
import sys
input = sys.stdin.readline
sys.setrecursionlimit(10**7)


def dfs(a, b, c):
    global hn
    if l1[a][b+1] == 1 and (visited[a][b+1] == 0):
        visited[a][b+1] = c
        hn += 1
        dfs(a, b+1, c)
    if l1[a+1][b] == 1 and (visited[a+1][b] == 0):
        visited[a+1][b] = c
        hn += 1
        dfs(a+1, b, c)
    if l1[a-1][b] == 1 and (visited[a-1][b] == 0):
        visited[a-1][b] = c
        hn += 1
        dfs(a-1, b, c)
    if l1[a][b-1] == 1 and (visited[a][b-1] == 0):
        visited[a][b-1] = c
        hn += 1
        dfs(a, b-1, c)


n = int(input())
l1 = [[0]*(n+2) for _ in range(n+2)]
visited = [[0]*(n+2) for _ in range(n+2)]
cnt = 1

for i in range(n):
    tmp = input()
    for j in range(n):
        l1[i+1][j+1] = int(tmp[j])


hl = []
for i in range(1, n+1):
    for j in range(1, n+1):
        if l1[i][j] == 1 and visited[i][j] == 0:
            hn = 0
            dfs(i, j, cnt)
            if hn == 0:
                hn = 1
            hl.append(hn)
            cnt += 1

hl.sort()
print(len(hl))
for i in hl:
    print(i)

File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.java
import java.util.*;

public class Main {

    public static int[][] l1;
    public static int[][] visited;
    public static int n;
    public static int hn;
    public static int[][] d = {{0, 1}, {1, 0}, {-1, 0}, {0, -1}};

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        n = sc.nextInt();
        sc.nextLine(); // 버퍼 클리어

        l1 = new int[n + 2][n + 2];
        visited = new int[n + 2][n + 2];

        for (int i = 1; i <= n; i++) {
            String tmp = sc.nextLine();
            for (int j = 0; j < n; j++) {
                l1[i][j + 1] = tmp.charAt(j) - '0';
            }
        }

        List<Integer> hl = new ArrayList<>();
        int cnt = 1;

        for (int i = 1; i <= n; i++) {
            for (int j = 1; j <= n; j++) {
                if (l1[i][j] == 1 && visited[i][j] == 0) {
                    hn = 0;
                    dfs(i, j, cnt);
                    if (hn == 0) {
                        hn = 1;
                    }
                    hl.add(hn);
                    cnt++;
                }
            }
        }

        Collections.sort(hl);
        System.out.println(hl.size());
        for (int i : hl) {
            System.out.println(i);
        }

        sc.close();
    }

    static void dfs(int x, int y, int c) {
        visited[x][y] = c;
        hn++;

        for (int[] direction : d) {
            int nx = x + direction[0];
            int ny = y + direction[1];

            if (l1[nx][ny] == 1 && visited[nx][ny] == 0) {
                visited[nx][ny] = c;
                dfs(nx, ny, c);
            }
        }
    }
}
